kinesis/kinesis.go

```diff
@@ -0,0 +1,311 @@
+func (outputPlugin *OutputPlugin) processAPIResponse(response *kinesis.PutRecordsOutput) error {
+    if aws.Int64Value(response.FailedRecordCount) > 0 {
+        // start timer if all records failed (no progress has been made)
+        if aws.Int64Value(response.FailedRecordCount) == int64(len(outputPlugin.records)) {
+            outputPlugin.timer.Start()
+            return fmt.Errorf("PutRecords request returned with no records successfully recieved")
+        }
+
+        logrus.Errorf("[kinesis %d] %d records failed to be delivered\n", outputPlugin.PluginID, aws.Int64Value(response.FailedRecordCount))
+        failedRecords := make([]*kinesis.PutRecordsRequestEntry, 0, aws.Int64Value(response.FailedRecordCount))
+        // try to resend failed records
+        for i, record := range response.Records {
+            if record.ErrorMessage != nil {
+                logrus.Debugf("[kinesis %d] Record failed to send with error: %s\n", outputPlugin.PluginID, aws.StringValue(record.ErrorMessage))
+                failedRecords = append(failedRecords, outputPlugin.records[i])
+            }
+            if aws.StringValue(record.ErrorCode) == kinesis.ErrCodeProvisionedThroughputExceededException {
+                // Backoff and Retry
+                outputPlugin.backoff.StartBackoff()
+            }
+        }
+
+        outputPlugin.records = outputPlugin.records[:0]
+        outputPlugin.records = append(outputPlugin.records, failedRecords...)
+        outputPlugin.dataLength = 0
+        for _, record := range outputPlugin.records {
+            outputPlugin.dataLength += len(record.Data)
+        }
+    } else {
+        //request fully succeeded
+        for i, record := range response.Records {
+            logrus.Debugf("[kinesis %d] record- %d:Shard ID: %s", outputPlugin.PluginID, i, aws.StringValue(record.ShardId))
+        }
+        outputPlugin.timer.Reset()
+        outputPlugin.backoff.Reset()
+        outputPlugin.records = outputPlugin.records[:0]
+        outputPlugin.dataLength = 0
+    }
+    return nil
+}
+
+//randomString generates a random string of length 8
+//it uses the math/rand library
+func (outputPlugin *OutputPlugin)randomString() string {
+    b := make([]byte, 16)
+    for i := range b {
+        b[i] = partitionKeyCharset[outputPlugin.seededRandom.Intn(len(partitionKeyCharset))]
+    }
+    return string(b)
+}
+
+//getPartitionKey returns the value for a given valid key
+//if the given key is emapty or invalid, it returns a random string
+func (outputPlugin *OutputPlugin) getPartitionKey(record map[interface{}]interface{}) string {
+    partitionKey := outputPlugin.partitionKey
+    if partitionKey != "" {
+        for k, v := range record {
+            dataKey := stringOrByteArray(k)
+            if(dataKey == partitionKey) {
+                value := stringOrByteArray(v)
+                if(value != "") {
+                    if len(value) > 256 {
+                        value = value[0:257]
```

Comment:
Pretty sure this should be value = value[0:256]

Reply:
already updated!